callouts/python/extproc/service/network_callout_server.py

```diff
@@ -0,0 +1,364 @@
+  def _loop_server(self) -> None:
+    """Loop server forever, calling shutdown will cause the server to stop."""
+
+    # We chose the main serving thread based on what server configuration
+    # was requested. Defaults to the health check thread.
+    if self._health_check_server:
+      logging.info("Health check server started.")
+      self._health_check_server.serve_forever()
+    else:
+      # If the only server requested is a grpc callout server, we wait on the grpc server.
+      self._callout_server.loop()
+
+  def shutdown(self) -> None:
+    """Tell the server to shutdown, ending all serving threads."""
+    if self._health_check_server:
+      self._health_check_server.shutdown()
+    if self._callout_server:
+      self._callout_server.stop()
+
+  def process(
+      self,
+      callout: ProcessingRequest,
+      context: ServicerContext,
+  ) -> ProcessingResponse:
+    """Process incoming network callouts.
+
+    Args:
+        callout: The incoming network callout.
+        context: Stream context on the callout.
+
+    Returns:
+        ProcessingResponse: A response for the incoming callout.
+    """
+    response = ProcessingResponse()
+
+    # Handle read data (client->server)
+    if callout.HasField('read_data'):
+      data = callout.read_data.data
+      end_of_stream = callout.read_data.end_of_stream
+
+      # Process the data through callback
+      processed_data, modified = self.on_read_data(data, end_of_stream, context)
+      
+      # Build response
+      response.read_data.data = processed_data
+      response.read_data.end_of_stream = end_of_stream
+      response.data_processing_status = (
+          ProcessingResponse.MODIFIED if modified
+          else ProcessingResponse.UNMODIFIED
+      )
+
+      # Check connection control
+      conn_status = self.should_close_connection(data, modified, context)
+
+    # Handle write data (server->client)
+    elif callout.HasField('write_data'):
+      data = callout.write_data.data
+      end_of_stream = callout.write_data.end_of_stream
+
+      # Process the data through callback
+      processed_data, modified = self.on_write_data(data, end_of_stream, context)
+
+      # Build response
+      response.write_data.data = processed_data
+      response.write_data.end_of_stream = end_of_stream
+      response.data_processing_status = (
+          ProcessingResponse.MODIFIED if modified
+          else ProcessingResponse.UNMODIFIED
+      )
+
+      # Check connection control
+      conn_status = self.should_close_connection(data, modified, context)
+
+    else:
+      logging.warning("Received request with no data")
+      return response
+
+    # Set connection status
+    response.connection_status = (
+        ProcessingResponse.CLOSE if conn_status
+        else ProcessingResponse.CONTINUE
+    )
+
+    return response
+
+  def on_read_data(
+      self,
+      data: bytes,
+      end_of_stream: bool,
+      context: ServicerContext,
+  ) -> Tuple[bytes, bool]:
```

Comment:
Consider using a namedtuple or dataclass with named fields `processed_data` and `modified` rather than a raw Tuple, to make the API more self-documenting and error-proof. See go/python-tips/019 for more info on these.

Reply:
done, dataclass makes sense.